content/post/performance-metrics-histogram.md

```diff
@@ -0,0 +1,245 @@
+
+### Cost Analysis
+
+- We have to store every single event in L.
+    - We have to send every single event to the monitoring hub so it can store it in L.  
+        __**Space cost: O(|L|)**__
+- We have to index L by time. 
+    - We can assume events arrive in order of completion, so L is naturally sorted.  
+      __**Time cost: 0**__
+- We have to select the window W from L.  
+    __**Space cost: O(|W|)**__  
+    __**Time cost: O(log(|W|))**__  
+- We have to sort W in order of latency values.
+    - Assuming best case performance of quicksort of O(nlogn)  
+        __**Time cost: O(|W|log(|W|))**__  
+
+__**Total time cost: O(|W|log(|W|)) + O(log|W|)**__  
+__**Total space cost: O(|L|) + O(|W|)**__
+
+## Using a Histogram
+
+Using a histogram, the following method is used: 
+
+1. We aggregate the histograms from a stream of events.
+    1. Local metrics are aggregated at the Metrics Receiver within each Wallaroo worker, and a histogram is produced per step per metrics period (typically 2 seconds).
+2. Once a metrics collection period elapses, the collection for that period is sent out to an external monitoring system.
+3. The histograms for each step are stored in a list L, sorted by time.
+4. The list for each step is stored in a hashtable H, keyed by step.
+  Note that this relationship is invertible and depends on how we prefer to query things.
+5. To get the latency histogram of a time window W, we choose all the histograms for a step and perform a bin-wise addition (since the bin ranges are identical across the histograms, this is safe). The result is the bin-wise sum of the histograms in the window, which is another histogram of the same type. We call this histogram S.
+    ![fig 1: performance metrics histogram](/images/post/performance-metrics-histogram/fig01_performance_metrics_histogram.png)
+
+6. We transform the histogram S into a cumulative histogram S', where each bin now includes the weight of all of the bins to the left of it (in addition to its own weight). 
+    ![fig 2: performance metrics cumulative histogram](/images/post/performance-metrics-histogram/fig02_performance_metrics_cumulative_histogram.png)
+
+7. We normalize the cumulative histogram to create a cumulative distribution D.
+    ![fig 3: performance metrics cumulative distribution](/images/post/performance-metrics-histogram/fig03_performance_metrics_cumulative_distribution.png)
+
+8. To answer whether the 99th- and 99.9th-percentile latencies are below our target, we simply compare the value in the appropriate bins to the target value. In the illustrated example, the fraction below 1s is 99.8%, and the fraction below 1ms is 24.67%, so it fails the SLA.
+   
+
+### Cost Analysis
+
+- Events are binned in a histogram with pre-determined bin value ranges
```

Comment:
It might be helpful to clarify where the cost is paid for each of these bullets (i.e. in Wallaroo vs in the external monitoring system).